index.js

```diff
@@ -1039,6 +1039,78 @@
       [Fn(a, $.Boolean), m(a), m(a)],
       Z.filterM);
 
+  //# takeWhile :: (Foldable f, Monoid (f a)) =>  (a -> Boolean) -> f a -> f a
+  //.
+  //. Returns the first elements for which the predicate returns true.
+  //.
+  //. ```javascript
+  //. > S.takeWhile(c => c < 'c', ['a', 'b', 'c', 'd', 'e'])
+  //. ['a', 'b']
+  //.
+  //. > S.takeWhile(c => c < 'a', ['a', 'b', 'c'])
+  //. []
+  //.
+  //. > S.takeWhile(S.K(true), S.Just(1))
+  //. Just(1)
+  //.
+  //. > S.takeWhile(S.K(false), S.Just(1))
+  //. Nothing
+  //. ```
+  function takeWhile(p, xs) {
+    if (Array.isArray(xs)) {
+      var result = [];
+      if (xs.length === 0) return result;
+      for (var i = 0, x; i < xs.length && p(x = xs[i]); i += 1) result.push(x);
+      return result;
+    }
+    var take = true;
+    function takeWhileReducer(xs, x) {
+      return take && p(x) ? append(x, xs) : (take = false, xs);
+    }
+    return Z.reduce(takeWhileReducer, Z.empty(xs.constructor), xs);
+  }
+  S.takeWhile =
+  def('takeWhile',
+      {f: [Z.Foldable, Z.Monoid]},
+      [Pred(a), f(a), f(a)],
+      takeWhile);
+
+  //# dropWhile :: (Foldable f, Monoid (f a)) =>  (a -> Boolean) -> f a -> f a
+  //.
+  //. Returns the elements following the first for which the predicate returns
+  //. false.
+  //.
+  //. ```javascript
+  //. > S.dropWhile(c => c < 'c', ['a', 'b', 'c', 'd', 'e'])
+  //. ['c', 'd', 'e']
+  //.
+  //. > S.dropWhile(c => c < 'a', ['a', 'b', 'c'])
+  //. ['a', 'b', 'c']
+  //.
+  //. > S.dropWhile(S.K(true), S.Just(1))
+  //. Nothing
+  //.
+  //. > S.dropWhile(S.K(false), S.Just(1))
+  //. Just(1)
+  //. ```
+  function dropWhile(p, xs) {
+    if (Array.isArray(xs)) {
+      if (xs.length === 0) return [];
+      for (var i = 0; i < xs.length && p(xs[i]); i += 1);
+      return xs.slice(i);
+    }
+    var drop = true;
+    function dropWhileReducer(xs, x) {
+      return drop && p(x) ? xs : (drop = false, append(x, xs));
```

Comment:
I think @cust0dian has the clearest statement using `done` and the assignment in the binary expression at the end. `done = true` is very explicit.